src/storage/src/hummock/iterator/forward_user.rs

```diff
@@ -26,6 +26,67 @@ use crate::hummock::value::HummockValue;
 use crate::hummock::HummockResult;
 use crate::monitor::StoreLocalStatistic;
 
+/// State transitions:
+/// Init       -> Ready:     The initial entry after seek/rewind can be exposed.
+/// Init       -> Next:      The initial entry after seek/rewind cannot be exposed.
+/// Init/Next  -> Invalid:   Iterator has been invalidated.
+/// Ready/Next -> Next:      Continue to call next on the inner iterator.
+/// Next       -> Ready:     The current entry can be exposed.
+enum IteratorState {
+    /// Initial state on iterator creation
+    Init,
+
+    /// Iterator is pointed to a valid key that can be exposed.
+    Ready,
+
+    /// Iterator is pointed to a key that should not be exposed. next() should be called.
+    Next,
+
+    /// Iterator is no longer valid. Reasons:
+    /// - Iterator is pointed to a out-of-range key.
+    /// - Iteraror meets EOF.
+    /// - The internal iterator is not valid.
+    /// - Error happens during iteration.
+    Invalid,
+}
+
+impl IteratorState {
+    #[inline]
+    fn is_next(&self) -> bool {
+        matches!(self, Self::Next)
+    }
+
+    #[inline]
+    fn is_ready(&self) -> bool {
+        matches!(self, Self::Ready)
+    }
+
+    #[inline]
+    fn is_init(&self) -> bool {
+        matches!(self, Self::Init)
+    }    
+
+    #[inline]
+    fn next(&mut self) {
+        *self = IteratorState::Next;
+    }
+
+    #[inline]
+    fn ready(&mut self) {
```

Comment:
When I read code, I always get lost in
1. iter.next()
2. state.next()
3. on_next()

Wouldn't it be easier to understand to explicitly use set_state(IteratorState) to change the state?  feel free to ignore me.

Reply:
This is a good point. I am also having a hard time on deciding the naming.

Reply:
It is indeed better to put `set_xxx` in the naming. Let me change it.